src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -397,148 +395,117 @@ public static IDataTransform Create(IHostEnvironment env, Arguments args, IDataV
             {
-                case DataKind.Text:
-                    return ComposeGetterOneCore(input.GetGetter<ReadOnlyMemory<char>>(srcCol), seed, mask);
                 case DataKind.U1:
-                    return ComposeGetterOneCore(input.GetGetter<byte>(srcCol), seed, mask);
+                    return MakeScalarHashGetter<byte, HashU1>(input, srcCol, seed, mask);
                 case DataKind.U2:
-                    return ComposeGetterOneCore(input.GetGetter<ushort>(srcCol), seed, mask);
+                    return MakeScalarHashGetter<ushort, HashU2>(input, srcCol, seed, mask);
                 case DataKind.U4:
-                    return ComposeGetterOneCore(input.GetGetter<uint>(srcCol), seed, mask);
+                    return MakeScalarHashGetter<uint, HashU4>(input, srcCol, seed, mask);
+                case DataKind.U8:
+                    return MakeScalarHashGetter<ulong, HashU8>(input, srcCol, seed, mask);
+                case DataKind.U16:
+                    return MakeScalarHashGetter<UInt128, HashU16>(input, srcCol, seed, mask);
+                case DataKind.I1:
+                    return MakeScalarHashGetter<sbyte, HashI1>(input, srcCol, seed, mask);
+                case DataKind.I2:
+                    return MakeScalarHashGetter<short, HashI2>(input, srcCol, seed, mask);
+                case DataKind.I4:
+                    return MakeScalarHashGetter<int, HashI4>(input, srcCol, seed, mask);
+                case DataKind.I8:
+                    return MakeScalarHashGetter<long, HashI8>(input, srcCol, seed, mask);
                 case DataKind.R4:
-                    return ComposeGetterOneCore(input.GetGetter<float>(srcCol), seed, mask);
+                    return MakeScalarHashGetter<float, HashFloat>(input, srcCol, seed, mask);
                 case DataKind.R8:
-                    return ComposeGetterOneCore(input.GetGetter<double>(srcCol), seed, mask);
+                    return MakeScalarHashGetter<double, HashDouble>(input, srcCol, seed, mask);
+                case DataKind.BL:
+                    return MakeScalarHashGetter<bool, HashBool>(input, srcCol, seed, mask);
                 default:
-                    Host.Assert(srcType.RawKind == DataKind.U8);
-                    return ComposeGetterOneCore(input.GetGetter<ulong>(srcCol), seed, mask);
+                    Host.Assert(srcType.RawKind == DataKind.Text);
+                    return MakeScalarHashGetter<ReadOnlyMemory<char>, HashText>(input, srcCol, seed, mask);
             }
         }
 
-        private ValueGetter<uint> ComposeGetterOneCore(ValueGetter<ReadOnlyMemory<char>> getSrc, uint seed, uint mask)
-        {
-            ReadOnlyMemory<char> src = default;
-            return
-                (ref uint dst) =>
-                {
-                    // REVIEW: Should we verify that the key value is within the advertised KeyCount?
-                    // Values greater than KeyCount should be treated as zeros.
```

Comment:
Does this need to be saved and moved up?

Reply:
I had missed that I had deleted this comment, but I'm not sure it's terribly valuable. In situations like this, while we *do* guard against anything that might reasonably cause, say, a crash (in which case I would argue we absolutely *must* handle this situation), if the idea is that garbage input will produce some possibly odd output, I'm not, quite honestly, too concerned.